packages/protocol/src/utils/SPID.ts

```diff
@@ -0,0 +1,202 @@
+ * Known keys
+ */
+export type SPIDObjectPartial = Partial<SPIDObject>
+
+/**
+ * SPID or String representing a SPID
+ * Object cases can be typechecked
+ * TODO: SPID string type safety
+ */
+export type SPIDLike = SPID | string | SPIDObject | { id: string, partition: number }
+
+/**
+ * Flexible input type
+ */
+export type SPIDLikePartial = SPIDLike | SPIDObjectPartial | Partial<{ id: string, partition: number }>
+
+class SPIDValidationError extends Error {
+    data: SPIDLikePartial
+    constructor(msg: string, data: SPIDLikePartial) {
+        super(format(msg, data))
+        this.data = data
+        if (Error.captureStackTrace) {
+            Error.captureStackTrace(this, this.constructor)
+        }
+    }
+}
+
+/**
+ * SPID - Stream Partition ID
+ */
+export default class SPID {
+    /** stream id */
+    public readonly id: string
+    /** stream partition */
+    public readonly partition: number
+    /** toString/fromString separator */
+    protected static readonly SEPARATOR = '#'
+
+    /** string key representing SPID */
+    public readonly key: string
+
+    /**
+     * @param id - stream id
+     * @param partition - stream partition
+     */
+    constructor(id: string, partition: number) {
+        this.id = id
+        this.partition = partition
+        this.validate()
+
+        // static cached values to prevent unnecessary runtime allocations
+        this.key = `${this.id}${SPID.SEPARATOR}${this.partition}`
+        // prevent all mutation
+        Object.freeze(this)
+    }
+
+    /**
+     * Throws if data is invalid.
+     */
+    private validate() {
+        if (typeof this.id !== 'string' || this.id.length === 0) {
+            throw new SPIDValidationError('SPID validation failed: id must be non-empty string: %o', {
+                id: this.id,
+                partition: this.partition
+            })
+        }
+
+        if (typeof this.partition !== 'number' || !Number.isSafeInteger(this.partition) || this.partition < 0) {
+            throw new SPIDValidationError('SPID validation failed: partition must be a safe integer >= 0: %o', {
+                id: this.id,
+                partition: this.partition
+            })
+        }
+    }
+
+    /**
+     * True iff other value is equivalent.
+     */
+    equals(other: SPIDLike): boolean {
+        // check if same instance
+        if (other === this) { return true }
+        let otherSpid: SPID
+        try {
+            otherSpid = SPID.from(other)
+        } catch (_err) {
+            // ignore error
+            // not equal if not valid
+            return false
+        }
+
+        // check key matches
+        return this.key === otherSpid.key
+    }
+
+    /**
+     * Convert SPIDLikePartial to SPIDObjectPartial
+     * i.e. normalizes various input types/shapes to { streamId?, streamPartition? }
+     * Note: does not throw on malformed input
+     */
+    static toSPIDObjectPartial(spidLike: SPIDLikePartial): SPIDObjectPartial {
```

Comment:
The`spidLike` is an optional parameter? We could remove the `!` cast at line 149.

Reply:
if the user calls `toSPIDObjectPartial` we probably want to force them to either verify input is valid or explicitly bang it.